tensorflow_probability/python/experimental/mcmc/particle_filter.py

```diff
@@ -354,130 +354,125 @@ def particle_filter(observations,
+            return smc_kernel.WeightedParticles(
+                particles=particles,
+                log_weights=log_weights + _compute_observation_log_weights(
+                    step + 1, particles, observations, observation_fn,
+                    num_transitions_per_observation=num_transitions_per_observation)
+            )
+
+
+    def sequential_monte_carlo(
+            initial_state,
+            propose_and_update_log_weights_fn,
+            condition_fn,
+            resample_fn,
+            resample_criterion_fn,
+            trace_fn
+    ):
+        kernel = smc_kernel.SequentialMonteCarlo(
+            propose_and_update_log_weights_fn=propose_and_update_log_weights_fn,
+            resample_fn=resample_fn,
+            resample_criterion_fn=resample_criterion_fn,
+            unbiased_gradients=unbiased_gradients)
+
+        # Use `trace_scan` rather than `sample_chain` directly because the latter
+        # would force us to trace the state history (with or without thinning),
+        # which is not always appropriate.
+        def seeded_one_step(seed_state_results, _):
+          seed, state, results = seed_state_results
+          one_step_seed, next_seed = samplers.split_seed(seed)
+          next_state, next_results = kernel.one_step(
+              state, results, seed=one_step_seed)
+          return next_seed, next_state, next_results
+
+
+        final_seed_state_result, traced_results = loop_util.trace_scan(
+            loop_fn=seeded_one_step,
+            initial_state=(loop_seed,
+                           initial_state,
+                           kernel.bootstrap_results(initial_state)),
+            elems=tf.ones([num_timesteps]),
+            trace_fn=lambda seed_state_results: trace_fn(*seed_state_results[1:]),
+            trace_criterion_fn=(
+                lambda seed_state_results: trace_criterion_fn(  # pylint: disable=g-long-lambda
+                    *seed_state_results[1:])),
+            static_trace_allocation_size=static_trace_allocation_size,
+            parallel_iterations=parallel_iterations,
+            condition_fn=condition_fn
+        )
+
+        if trace_criterion_fn is never_trace:
+          # Return results from just the final step.
+          traced_results = trace_fn(*final_seed_state_result[1:])
+
+        return traced_results
+
+
+    traced_results = sequential_monte_carlo(
+        # Weighted particles weighted also considering the observations.
+        initial_state=smc_kernel.WeightedParticles(
+            particles=initial_particles,
+            log_weights=initial_log_weights + _compute_observation_log_weights(
+                step=0,
+                particles=initial_particles,
+                observations=observations,
+                observation_fn=observation_fn)
+        ),
         propose_and_update_log_weights_fn=propose_and_update_log_weights_fn,
+        condition_fn=lambda step, state, num_traced, trace: step < len(observations),
```

Comment:
Because `observations` is a Tensor, we need to use a TensorFlow operation here to get the length.  E.g., `tf.shape(observations)[0]` will return a Tensor whose value is the first dimension of the shape of `observations`.

We expect this code to be faster to run (and to XLA compile) if we can use a constant ("statically-known") value instead of a computed Tensor for this bound.  We have utility function in TFP that will return a statically-known value when possible and will fall back to returning a Tensor when necessary.  For this particular case (getting the length of a Tensor), we prefer to use `ps.size0(observations)` over `tf.shape(observations)[0]`.